[ETHEREUM-CONTRACTS] Adding foundry toga tests

shreyaspapi opened the pull request:
Test -

- [x] Remaining `#11 - #13` 
- [x] Remaining `#14 - #15`

From https://github.com/superfluid-finance/protocol-monorepo/blob/dev/packages/ethereum-contracts/test/contracts/utils/TOGA.test.ts

shreyaspapi committed:
Adding foundry toga tests

shreyaspapi committed:
Adding more tests

shreyaspapi committed:
Refactor

hellwolf commented:
@d10r would you be interested in reviewing it too?

hellwolf commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,483 @@
+            abi.encode(EXIT_RATE_1)
+        );
+
+        assertTrue(toga.getCurrentPIC(superToken) == alice);
```
missing `vm.expectEmit` testing that's present in typescript?

hellwolf replied:
Annoyingly, would need to copy the event signature in this contract

shreyaspapi replied:
Looking into it 👀

shreyaspapi replied:
Added `vm.expectEmit` to test all the events.

hellwolf commented:
Perhaps we can add a fuzzing test case now that foundry is being used?

d10r commented:
hi @shreyaspapi , thx for that!

A few observations:
* using foundry, we could simplify the testing code by taking advantage of its fuzzing tests. Those are especially useful for figuring out edge cases. I added an example in https://github.com/superfluid-finance/protocol-monorepo/commit/7137c9488d6f0a795f8eb20ea9b82963bbefd059 - feel free to copy'n paste that & iterate on it
* `_collectRewards` was a kind of workaround in order to avoid triggering the ERC777 callback. We disabled calling those when using `ERC20.transfer()` later, thus this isn't needed anymore. If it remained like this, you should also migrate over the comments in the JS files, otherwise that context gets lost. But I suggest to instead remove that function altogether and instead use foundry's [deal](https://book.getfoundry.sh/cheatcodes/deal?highlight=deal#deal).
* you can also use `deal` instead of mint
* there's an edge case not covered by the tests (also missing in the JS tests): if the toga bond gets negative (possible if a token goes insolvent), all bids fail. You may try to add a fuzzing test case which uncovers that

shreyaspapi committed:
Fixing tests

shreyaspapi commented:
@hellwolf @d10r

Thanks for the initial review. I have made some changes as suggested.

- Added fuzzing tests using Foundry to simplify testing and uncover edge cases.
- Removed the `_collectRewards` function and replaced it with Foundry's `deal` functionality.

> - there's an edge case not covered by the tests (also missing in the JS tests): if the toga bond gets negative (possible if a token goes insolvent), all bids fail. You may try to add a fuzzing test case which uncovers that

Regarding the missing edge case mentioned about the toga bond going negative, I believe that if the stream of toga -> PIC gets liquidated, the bond value goes to 0. Please clarify if there's something else I need to consider, can you explain a bit more about `(possible if a token goes insolvent)`.

Please review the changes and let me know if there's anything else to address other then toga going to negative.

shreyaspapi committed:
Merge branch 'dev' into toga-test

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+
+    TOGA internal toga;
+
+    uint256 internal constant MIN_BOND_DURATION = 3600 * 24 * 7; // 7 days
```
you can use `sfDeployer.DEFAULT_TOGA_MIN_BOND_DURATION` to retrieve the canonical min bond duration.
also, you can use `7 days` or `1 week` instead of `3600 * 24 * 7` in solidity

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+// SPDX-License-Identifier: AGPLv3
```
please do a forge fmt on this file: 
`forge fmt test/foundry/utils/TOGA.t.sol`

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+     * @param expectedNetFlow The expected net flow.
+     */
+    function _assetNetflow(
+        ISuperToken token,
```
please use `superToken_` instead of `token` as this shadows an existing declaration in storage.

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+    }
+
+    function _deleteFlow(
+        ISuperToken token,
```
please refer to `FoundrySuperfluidTester.sol`, there are helper functions there which do a lot more assertions.
see `_helperDeleteFlow`

0xdavinchee replied:
ditto w/ token comment from above.

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+        vm.stopPrank();
+    }
+
+    function _startStream(
```
same as the above with `_deleteFlow`

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+
+    function _emptyAccount(
+        address account,
+        ISuperToken token
```
ditto w/ token comment from above.

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+    }
+
+    function _transferToken(
+        IERC20 token,
```
ditto w/ token comment from above. you can use `token_` here

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+// SPDX-License-Identifier: AGPLv3
+pragma solidity 0.8.19;
```
the tests are 100% coverage.
please delete the `TOGA.test.ts` file and also remove the line in `testsuites/all-contracts.ts` which references this file to be run for testing.

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+     * @param account The address of the account to check.
+     * @param expectedNetFlow The expected net flow.
+     */
+    function _assetNetflow(
```
You might not need this function with the helper functions after

0xdavinchee replied:
if it is needed, move this to `FoundrySuperfluidTester.sol`, also a minor typo: `_assertNetFlow` => `assertNetFlow`

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+     */
+    function _sendPICBid(
+        address sender,
+        ISuperToken token,
```
ditto w/ token comment from above (use `superToken_`)

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+
+    function _sendPICBid(
+        address sender,
+        ISuperToken token,
```
ditto w/ token comment from above (use `superToken_`)

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+        vm.stopPrank();
+    }
+
+    function _shouldDefaultExitRate(
```
I see that this is a reimplementation `getDefaultExitRateFor`, please use `toga.getDefaultExitRateFor()` instead.

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+        return capToInt96(int256(bondAmount / (MIN_BOND_DURATION * 4)));
+    }
+
+    function capToInt96(int256 value) internal pure returns(int96) {
```
same as above, best not to duplicate

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+        return value < type(int96).max ? int96(value) : type(int96).max;
+    }
+
+    function _shouldMaxExitRate(
```
same comment as above

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+     * @dev Tests that Alice becomes the PIC.
+     */
+    function testAliceBecomesPIC(uint256 bond, int96 exitRate) public {
+        assertTrue(address(0) == toga.getCurrentPIC(superToken));
```
you can alternatively use assertEq(), this makes it more immediately clear that you are asserting that two values are equal than having to do an extra mental step of seeing what is true

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+         // satisfy exitRate constraints of the TOGA
+         vm.assume(exitRate <= toga.getMaxExitRateFor(superToken, bond));
+         // the clipped CFA deposit needs to fit into 64 bits - since that is flowrate multiplied by
+         // liquidation perdion, 14 bits are added for 14400 seconds, so we can't use the full 96 bits
```
I would look at some of the assumes and you can move them out into an `_assume` function, see `_assumeValidFlowRate` in `FoundrySuperfluidTester.sol`.

Also a small typo: `perdion` => `period`

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+            superToken
+        );
+
+        assertTrue(pic == alice);
```
use `assertEq` when possible and add an error string at the end of the assert functions for easier debugging. see some of the other test files for the conventions:
`"testAliceBecomesPIC: Alice is PIC"`
The convention is: `"{NAME_OF_FUNCTION}: {BRIEF_DESCRIPTION_OF_ASSERTION}"`

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+        // Send PIC bid from Alice
+        _sendPICBid(alice, superToken, aliceBond, 0);
+
+        vm.startPrank(alice);
```
this prank may not be needed?

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+
+        vm.stopPrank();
+
+        vm.expectRevert("TOGA: bid too low");
```
please move this expectRevert function into its own test:
`function testRevertIfBidTooLow(...)`
You can do a global search for: `testRevertIf` to see examples of this

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+            keccak256("TOGAv2")
+        );
+
+        assertTrue(
```
can use assertEq here too

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+        );
+    }
+
+    function testEnforceMinExitRateLimit() public {
```
just a convention thing, but use `testRevertIf__` naming convention and try to keep the tests as small as possible, that is if this test is expecting a revert, it should only be doing that.

So this test could be split in two, if the second part where you send the pic bid to get a 0 net flow rate is not tested elsewhere already.

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+                toga.getDefaultExitRateFor(superToken, bond)
+        );
+        // the default exit rate needs to be equal or greater than the max exit rate
+        assertTrue(
```
you can also use `assertGe()` :)

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+        );
+    }
+
+    function testCannotBecomePICWithBidSmallerThanCurrentPICBond(uint256 bond, uint256 smallerBond) public {
```
please use the `testRevertIf` naming

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+        );
+
+        assertTrue(pic == alice);
+        assertTrue(bond == bond);
```
I would change the name of the param here to `bond_`

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+// SPDX-License-Identifier: AGPLv3
+pragma solidity 0.8.19;
+
```
I'd highly recommend `Solidity Visual Developer` if you're using VSCode as well :)

0xdavinchee commented on packages/ethereum-contracts/test/foundry/utils/TOGA.t.sol:
```diff
@@ -0,0 +1,553 @@
+        (, uint256 bond2, ) = toga.getCurrentPICInfo(superToken);
+        assertTrue(bond2 >= 1e12);
+
+        vm.startPrank(alice);
```
the changing of the exit rate likely can be placed in a helper function where you assertNetFlow after.

0xdavinchee reviewed (commented):
gm! 

This looks very good overall, the logic is being tested well and you have 100% coverage for the TOGA, so props! We are very grateful that you've written these tests.

Almost all of my comments are regarding convention and the way of writing tests at Superfluid, you can review the [README](https://github.com/superfluid-finance/protocol-monorepo/blob/dev/packages/ethereum-contracts/README.md) in ethereum-contracts to get a better idea of the how and why.

Some high level comments, I've referenced some examples in the code with comments as well:
- aim to not duplicate logic as this leads to maintenance overhead and possible bugs down the line
- utilize existing helper functions in `FoundrySuperfluidTester.sol` when possible
- be careful with shadowed declarations, the plugin I recommended will make this visible immediately
- use `assertEq`, `assertGe`, etc. when possible and use `assertTrue` when the other options aren't available.
- there are `assume's` for the different fuzzing tests which are repeated, try to see which ones can be placed in an `_assumeXYZ` function, this follows from the duplication point above
- keep the tests specific to testing one thing, if it's supposed to test a passing case, only test that, if it's supposed to revert, only test that, you may need several steps to test certain cases (e.g. testMultiplePICsInParallel), but the reverts in there can be taken out into their own revert tests
  - use `testRevertIf` naming convention

hellwolf commented:
should also remove the original test file then we are ready to merge.

hellwolf reviewed (approved):
remove js test then we should be good!

0xdavinchee committed:
cleanup

- remove toga.test.ts file
- fix up depositDelta in FoundrySuperfluidTester.sol
- address review comments in TOGA.t.sol

shreyaspapi commented:
@0xdavinchee I see you have made a commit, I also had some uncommitted changes 😕, that addresses the issues.

0xdavinchee commented:
@shreyaspapi oh okay, sorry about that! I wanted to help you with some of the changes because I requested a lot of different things.

0xdavinchee committed:
more cleanup

0xdavinchee committed:
Merge branch 'dev' into toga-test

0xdavinchee committed:
fix toga foundry tests

0xdavinchee committed:
rename test contract

0xdavinchee committed:
toga tests cleanup

0xdavinchee committed:
Merge branch 'dev' into toga-test

hellwolf committed:
Merge branch 'dev' into toga-test

hellwolf committed:
Merge branch 'dev' into toga-test
